### PR TITLE
Div and Alice: OO-Ride-Share

### PR DESCRIPTION
**Pull request by alicehsiao**
# OO Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | In Load Drivers, we decided to model it after the other two provided methods (Load Trips and Load Users). In addition, we added the piece where we replace passengers with users in this method. We initially created a new method to do this, but talked to other students in our class and this seemed like the neatest way.
Describe any examples of composition that you encountered in this project, if any | TripDispatcher creates instances of User, Driver, and Trips in the load data methods. 
Describe the relationship between `User` and `Driver`| 'Driver' inherits from 'User', and thus is a sub class of User. Both are contained in the RideShare module. 
Describe a _nominal test_ that you wrote for this assignment. | We wrote a test for Driver#total_revenue that accurately calculates the total revenue for given driver. 
Describe an _edge case test_ that you wrote for this assignment | We wrote a test for Driver#average_rating and checked that it returned 0 if there were no driven trips.
Describe a concept that you/your pair gained more clarity on as you worked on this assignment | Our understanding of how 'super' worked with inheritance was improved. We were also able to review concepts of composition when trying to access attributes of User/Driver/Trip instances in different files.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We talked about trying to find the right balance in our pair programming experience with giving and receiving ideas. We also talked about the differences in our communication styles when working with partners.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      else
+        raise ArgumentError, 'Trip is not provided'
+      end
+    end
+
+    def average_rating
+      result = 0
+      if @driven_trips == []
+        return 0
+      else
```
Comment (dHelmgren) [inline lib/driver.rb:35]: See my comment in User. You have already created a result of 0 to pass back, so why bother with the `if else` ya know?

File: lib/user.rb [hunk 2]
```diff
@@ -11,10 +14,31 @@ def initialize(input)
...
       @trips = input[:trips].nil? ? [] : input[:trips]
+
     end
 
     def add_trip(trip)
       @trips << trip
     end
+
+    def net_expenditures
+      if @trips == 0
```
Comment (dHelmgren) [inline lib/user.rb:25]: While this works, I wouldn't actually implement it this way. `if @trips == 0` isn't the question I think you are trying to answer, what you are trying to ask is whether or not `@trips` has one or more elements, which would be `if @trips.length == 0`

All of which is somewhat besides the point, as you set `total_spent` to zero at the beginning of the `.each` loop.

File: specs/trip_spec.rb [hunk 3]
```diff
@@ -27,8 +30,29 @@
...
 
+    it "checks to make sure start time is before end time" do
+      start_time = Time.parse('2015-05-20T12:14:00+00:00')
+      end_time = start_time + 25 * 60 # 25 minutes
+      my_trip = {
+        id: 8,
+        passenger: RideShare::User.new(id: 1,
+                                       name: "Ada",
+                                       phone: "412-432-7640"),
+        start_time: end_time,
```
Comment (dHelmgren) [inline specs/trip_spec.rb:41]: This is a confusing way to set up this test. Keeping the hash assignment consistent with the variable name you are using is an important convention.

File: lib/driver.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+          total_revenue += 0.8 * (trip.cost - 1.65) if trip.cost != nil
+        end
+      end
+      
+      return total_revenue.round(2)
+    end
+
+    def net_expenditures
+      made_as_driver = total_revenue
+      spent_as_passenger = super
```
Comment (dHelmgren) [inline lib/driver.rb:57]: So, we haven't talked about this a lot, but you can treat a method as a variable on its own as long as its return value responds properly to the operation. For example, here, `super` and `total_revenue` will respond to the `-` operator because they return floats. 

You can write this entire method in one line as `return (super - total_revenue).round(2)` if you so desire.

File: specs/driver_spec.rb [hunk 5]
```diff
@@ -1,104 +1,146 @@
...
-      trip2 = RideShare::Trip.new(id: 8, driver: @driver, passenger: nil,
-                                  date: Time.parse("2016-08-08"), rating: 1)
-      @driver.add_trip(trip2)
+    describe "total_revenue" do
+      it "accurately calculates the revenue for driver" do
+        @driver.add_driven_trip(@trip_3)
 
-      expect(@driver.average_rating).must_be_close_to (5.0 + 1.0) / 2.0, 0.01
+        expect(@driver.total_revenue).must_equal 48.60
+      end
```
Comment (dHelmgren) [inline specs/driver_spec.rb:129]: What if the driver has no trips?

File: lib/trip_dispatcher.rb [hunk 6]
```diff
@@ -60,6 +97,50 @@ def find_passenger(id)
...
+         new_available_drivers.delete(instance)
+        end
+      end
+
+      if new_available_drivers == []
+        raise ArgumentError.new("No Available Drivers. Try later.")
+      else
+        driver = new_available_drivers[0]
+        new_available_drivers.each_with_index do |instance, index|
+          if instance.driven_trips == []
```
Comment (dHelmgren) [inline lib/trip_dispatcher.rb:121]: So, what I am seeing is that we were looking for a driver who hasn't made any driven trips, who is farthest back in the array (closest to `new_available_drivers[-1]`). Is this what you meant to do?

File: specs/trip_dispatch_spec.rb [hunk 7]
```diff
@@ -111,4 +92,95 @@
...
+    end
+
+    it "raises an ArgumentError if there are no available drivers" do
+      @dispatcher.request_trip(6)
+      @dispatcher.request_trip(3)
+
+      expect{@dispatcher.request_trip(1)}.must_raise ArgumentError
+    end
+
+    it "checks that drivers are not driving themselves" do
```
Comment (dHelmgren) [inline specs/trip_dispatch_spec.rb:132]: This is a good test to do, but I don't think it tests what you want it to test. You should test it in a circumstance when the ONLY available driver is the passenger that you are trying to drive.

File: specs/user_spec.rb [hunk 8]
```diff
@@ -59,4 +58,40 @@
...
+                                 end_time: Time.parse("2016-08-09"), cost: 5.25,
+                                 rating: 5)
+      trip_2 = RideShare::Trip.new(id: 8, driver: nil, passenger: @user,
+                                  start_time: Time.parse("2016-08-07"),
+                                  end_time: Time.parse("2016-08-10"), cost: 10.84,
+                                  rating: 5)
+      trip_3 = RideShare::Trip.new(id: 8, driver: nil, passenger: @user,
+                                  start_time: Time.parse("2016-08-01"),
+                                  end_time: Time.parse("2016-08-03"), cost: 35.20,
+                                  rating: 5)
```
Comment (dHelmgren) [inline specs/user_spec.rb:77]: Looks like you didn't come back to test that your stuff from wave 1 works with your incomplete trips from Wave 3!

File: specs/trip_dispatch_spec.rb [hunk 9]
```diff
@@ -111,4 +92,95 @@
...
+      expect(average_hourly_rate).must_equal 31.28
+    end
+
+    it "will ensure that the driver who is selected has no trips" do
+      new_trip = @dispatcher_2.request_trip(3)
+      driver = new_trip.driver
+
+      expect(driver.driven_trips.length).must_equal 1
+      expect(driver.id).must_equal 8
+    end
```
Comment (dHelmgren) [inline specs/trip_dispatch_spec.rb:184]: What if there is no available driver? This is an important case, and is exactly the sort of situation where software will do something unexpected. You might set up the test as follows:

```ruby
it 'raises an error if no driver is available' do
  @dispatcher.drivers.each do |driver|
    # This may require an attr_writer in Driver
    driver.status = :UNAVAILABLE
  end
  proc {
    @dispatcher.request_trip(2)
  }.must_raise StandardError
end
```

**Comment by dHelmgren** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Sure did!
Answer comprehension questions | Good answers here!
**Wave 1** |
Appropriate use of Ruby's `Date`| Yes!
`Trip` has a helper method to calculate duration| Check
`User` (passenger) has a method to calculate total cost of all trips| Check!
Tests for wave 1 | The tests cover the ground well, but see my comment. When you create the test data and assign start to end in your hash, it can be hard to read/easy to miss
**Wave 2** |
`Driver` inherits from `User` | Check
`Driver` has `add_driven_trip` method | check!
`Driver` has method to calculate average rating | Check! 
`Driver` has method to calculate net expenditures and it uses `super` | Check 
`Driver` has a method to calculate total revenue | Check!
Tests for wave 2 | Your tests for wave 2 are good!
**Wave 3** |
`TripDispatcher` has a new method to create trips | Check!
creating a trip in `TripDispatcher` relies on methods in `Driver` and `User` (passenger) to modify their own attributes | Yeah!
Complex logic was correctly implemented | Looks good!
Tests for `request_trip` | A few small hiccups, see my comments! 
Methods from wave 1 and 2 handle incomplete trips | They do!
Tests for wave 1 and 2 methods with incomplete trips |  I think you're missing some edge cases here.
**Wave 4 (Optional)** |
`TripDispatcher` now assigns trips to either the newest driver (no trips), or the driver who has not driven in the longest time | Looks like you implemented this half-way. See my comment.
Appropriate helper methods were made to help with complex logic | 
Tests for wave 4 | 
**Overall** | I think that you have a really solid foundation on a tough project. Make sure you are testing more edge cases, and that you revise your tests as you add functionality!

**Review by dHelmgren — COMMENTED** [review]
Big Takeway: Keep an eye on how tests need to change as you go through a project!
